lib/pbench/agent/tool_data_sink.py

```diff
@@ -741,89 +996,193 @@ def _wait_for_all_data(self):
+                    section = "pbench"
+                    try:
+                        mdlog.add_section(section)
+                    except DuplicateSectionError:
+                        pass
+                    iterations_l = iterations_val.strip().split()
+                    iterations_str = ", ".join(iterations_l)
+                    mdlog.set(section, "iterations", iterations_str)
+                # Write out the final meta data contents.
+                with (mdlog_name).open("w") as fp:
+                    mdlog.write(fp)
+
+            self._from_client_chan.close()
+            self._from_tms_chan.close()
+            self._to_logging_chan.unsubscribe()
+            return
+
         directory = Path(directory_str)
         if not directory.is_dir():
             self.logger.error(
-                "state change to '%s' with non-existent directory, '%s'",
-                data["action"],
-                directory,
+                "action '%s' with non-existent directory, '%s'", action, directory,
             )
-            raise self.Terminate("Terminate bottle server w/ error")
+            self._send_client_status(action, "invalid directory")
+            return
         try:
             # Check that "directory" has a prefix of self.benchmark_run_dir
             directory.relative_to(self.benchmark_run_dir)
         except ValueError:
             self.logger.error(
-                "state change to '%s' with invalid directory,"
+                "action '%s' with invalid directory,"
                 " '%s' (not a sub-directory of '%s')",
-                data["action"],
+                action,
                 directory,
                 self.benchmark_run_dir,
             )
-            raise self.Terminate("Terminate bottle server w/ error")
-        else:
-            self.directory = directory
-        # The remote tool meisters will be hashing the directory argument this
-        # way when invoking the PUT method.  They just consider the directory
-        # argument to be an opaque context.  We, the tool data sink, write the
-        # data we receive to that directory, but expect them to provide the
-        # opaque context in the URL for the PUT method.
-        directory_bytes = directory_str.encode("utf-8")
-        self.data_ctx = hashlib.md5(directory_bytes).hexdigest()
-
-        # Transition to "send" state should reset self._tm_tracking
+            self._send_client_status(action, "directory not a prefix of run directory")
+            return
+
         with self._lock:
-            if self.state == "init":
+            # Handle all actions underneath the lock for consistency.
+            self.action = action
+            if action == "init":
+                # To be safe, clear the data context to catch bad PUTs
+                self.data_ctx = None
+                # Forward to TMs
+                ret_val = self._forward_tms_and_wait(data)
```

Comment:
@Maxusmusti this is the fix for PCP tools not working; we forward to the Tool Meisters first, and then handle the PCP pmlogger collectors.